### PR TITLE
Fix bugs reported by mercury#4

### PR DESCRIPTION
**Pull request by rubio41**
1. change the new DateObject to DateObject.SetValue extension method which will check if the date object is valid or not to avoid argument exceptions.
2. fix the bug in "to" as a connector to restrict the exact match
3. restrict the month regex and day regex to end with '\b'

### CODE REVIEW

File: Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 1]
```diff
@@ -23,11 +23,11 @@ public static class DateTimeDefinitions
 		public const string NextPrefixRegex = @"(next|upcoming)\b";
 		public const string PastPrefixRegex = @"(last|past|previous)\b";
 		public const string ThisPrefixRegex = @"(this|current|over the)\b";
-		public const string DayRegex = @"(the\s*)?(?<day>01|02|03|04|05|06|07|08|09|10th|10|11th|11st|11|12nd|12th|12|13rd|13th|13|14th|14|15th|15|16th|16|17th|17|18th|18|19th|19|1st|1|20th|20|21st|21|22nd|22|23rd|23|24th|24|25th|25|26th|26|27th|27|28th|28|29th|29|2nd|2|30th|30|31st|31|3rd|3|4th|4|5th|5|6th|6|7th|7|8th|8|9th|9)";
-		public const string MonthNumRegex = @"(?<month>01|02|03|04|05|06|07|08|09|10|11|12|1|2|3|4|5|6|7|8|9)";
+		public const string DayRegex = @"(the\s*)?(?<day>01|02|03|04|05|06|07|08|09|10th|10|11th|11st|11|12nd|12th|12|13rd|13th|13|14th|14|15th|15|16th|16|17th|17|18th|18|19th|19|1st|1|20th|20|21st|21|22nd|22|23rd|23|24th|24|25th|25|26th|26|27th|27|28th|28|29th|29|2nd|2|30th|30|31st|31|3rd|3|4th|4|5th|5|6th|6|7th|7|8th|8|9th|9)(?=\b|t)";
+		public const string MonthNumRegex = @"(?<month>01|02|03|04|05|06|07|08|09|10|11|12|1|2|3|4|5|6|7|8|9)\b";
```
Comment (tellarin) [inline Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Can you check if Spanish also needs this fix?

File: Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 2]
```diff
@@ -23,11 +23,11 @@ public static class DateTimeDefinitions
...
-		public const string MonthNumRegex = @"(?<month>01|02|03|04|05|06|07|08|09|10|11|12|1|2|3|4|5|6|7|8|9)";
+		public const string DayRegex = @"(the\s*)?(?<day>01|02|03|04|05|06|07|08|09|10th|10|11th|11st|11|12nd|12th|12|13rd|13th|13|14th|14|15th|15|16th|16|17th|17|18th|18|19th|19|1st|1|20th|20|21st|21|22nd|22|23rd|23|24th|24|25th|25|26th|26|27th|27|28th|28|29th|29|2nd|2|30th|30|31st|31|3rd|3|4th|4|5th|5|6th|6|7th|7|8th|8|9th|9)(?=\b|t)";
+		public const string MonthNumRegex = @"(?<month>01|02|03|04|05|06|07|08|09|10|11|12|1|2|3|4|5|6|7|8|9)\b";
 		public const string PeriodYearRegex = @"\b(?<year>19\d{2}|20\d{2})\b";
-		public const string WeekDayRegex = @"(?<weekday>Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Mon|Tue|Wedn|Weds|Wed|Thurs|Thur|Thu|Fri|Sat|Sun)";
-		public static readonly string RelativeMonthRegex = $@"(?<relmonth>{RelativeRegex}\s+month)";
+		public const string WeekDayRegex = @"(?<weekday>Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Mon|Tue|Wedn|Weds|Wed|Thurs|Thur|Thu|Fri|Sat|Sun)\b";
+		public static readonly string RelativeMonthRegex = $@"(?<relmonth>{RelativeRegex}\s+month)\b";
```
Comment (tellarin) [inline Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Same.

File: Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/HolidayParserChs.cs [hunk 3]
```diff
@@ -392,31 +392,31 @@ private static int ConvertChineseToInteger(string yearChsStr)
...
-
-        internal static readonly DateObject midautumnday = new DateObject(now.Year, 8, 15);
-        internal static readonly DateObject springday = new DateObject(now.Year, 1, 1);
-        internal static readonly DateObject chuxiday = new DateObject(now.Year, 1, 1).AddDays(-1);
-
-        internal static readonly DateObject lanternday = new DateObject(now.Year, 1, 15);
-        internal static readonly DateObject qingmingday = new DateObject(now.Year, 4, 4);
-        internal static readonly DateObject dragonboatday = new DateObject(now.Year, 5, 5);
-        internal static readonly DateObject chongyangday = new DateObject(now.Year, 9, 9);
+        internal static readonly DateObject yuandan = DateObject.MinValue.SetValue(now.Year, 1, 1);
```
Comment (tellarin) [inline Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/HolidayParserChs.cs]: We should keep the old creation in constants. Only use your new method in DateObjects that are created from user provided tokens.

File: Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/TimePeriodParserChs.cs [hunk 4]
```diff
@@ -1,237 +1,237 @@
...
+            int day = refTime.Day,
+                month = refTime.Month,
+                year = refTime.Year;
+
+            //determine if the right side time is smaller than the left side, if yes, add one day
+            int hour = leftResult.Hour > 0 ? leftResult.Hour : 0,
+                min = leftResult.Minute > 0 ? leftResult.Minute : 0,
+                second = leftResult.Second > 0 ? leftResult.Second : 0;
+
+            var leftTime = DateObject.MinValue.SetValue(year, month, day, hour, min, second);
```
Comment (tellarin) [inline Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/TimePeriodParserChs.cs]: We could rename this to SafeCreateFromValues or something like it.

File: Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 1 above)
Comment (rubio41) [inline Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: do the same change in Spanish

File: Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 2 above)
Comment (rubio41) [inline Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: do the same change in Spanish

**Pull request merged**